### ISSUE TITLE
Random Seed vs. Repeated Experiments?

### ISSUE THREAD

**Issue by mlindauer** [berndbischl/coseal-algsel-benchmark-repo#78]
Hi,

I encountered an issue in my tool 
that could be also an issue here.
Some models are really performance-sensitive to the random seed (e.g., libSVM).
Do we fix the random seed to an arbitrary value?
Or do we have to perform several runs?

How is it normally handled in ML?

Cheers,
Marius

**Comment by berndbischl** [comment]
Please be more specific in describing the problem and explain what you saw when.

Also: libSVM is very likely a deterministic algorithm. 
Be more specific here, too!
Which SVM, kernel, hyperparameters?
E.g. in R (when interfacing the libsvm library), per default the RBF kernel is selected. The optimizer for the resulting problem is deterministic. BUT there is a default STOCHASTIC heuristic to set the kernel width, depending on the training data.

**Comment by mlindauer** [comment]
ok, here is what I did:
I learned a probability estimating SVC model with libSVM and set the kernel to rbf, and also defined a gamma and C. Depending on the used random seed, the model learned on the same training data differed.

But also other models, such as Random Forests, uses some randomness.
Therefore, the seed would be important to replicate results, right?
So it should be an issue independent of the used model as long as random decision are used.

**Comment by berndbischl** [comment]
> I learned a probability estimating SVC model with libSVM and set the kernel to rbf, 
> and also defined a gamma and C. Depending on the used random seed, the model learned on
> the same training data differed.

See now I can answer. Prob estimation in libsvm uses Platt's heuristic.
See page 30 to educate yourself about that:
http://www.csie.ntu.edu.tw/~cjlin/papers/libsvm.pdf

You will also see that that uses internally a cross-validation = stochastic behavior. 
Also note the comment: 

"we conduct five-fold cross-validation to obtain  decision  values  before  minimizing  the  negative  log  likelihood.   Note  that  if some classes contain five or even fewer data points, the resulting model may not be good"

If you do vanilla classification, you might see exactly the instability due to that fact, as the class "best solver" might have levels with very small sizes.

**Comment by mlindauer** [comment]
thanks for explaining this for the this specific case.
What do we do in the general case when a model is trained in a stochastic way?

**Comment by berndbischl** [comment]
> But also other models, such as Random Forests, uses some randomness.
> Therefore, the seed would be important to replicate results, right?
> So it should be an issue independent of the used model as long as random decision are used.

You are in general correct but this is a very difficult point. I need some time to explain this. And after that you will potentially we disappointed that what you want is not really possible.

**Comment by mlindauer** [comment]
ok, so you are aware of the problem (also in our setting),
and you will explain it to me during a beer (or wine) at the next coseal workshop ;-)

if yes, we can close the issue.

**Comment by berndbischl** [comment]
We are talking about randomness and instability of benchmark results and a fear of non-reproducebility, when such randomness is high.

The are 2 sources of randomness when we produce our results:

a) the CV is stochastic 
b) the model can be stochastic, e.g. a RANDOM forest

Usually, when people start talking about this, they want to artificially "fix" the randomness to some deterministic behavior so one can reproduce results to the last digit, when running code again. 
One can use seeding. We also do something comparable when we predefine the CV splits. 
(It has another advantage of reducing variance when comparing models on the same splits, and not different ones).

Now there are reasons why this is "nice to have". But does it solve the real problem? Lets just consider a) for the moment and our solution of defining the CV. If another random CV now completely changes the benchmark results in a drastic way, we are still fucked. One could even argue that our solution of fixing the CV splits makes it WORSE, because it tends to HIDE that fact from users, as everybody just uses our CV splits and nobody sees the true variance. 
(Note that our CV splits are completely arbitrary and not more or less "realistic" than any other 10CV splitting).

Now what is the solution to that:

(*) Perform repetitions (of the CV). This enables us to see that variance. We should display it. If it is high, we should discuss it. We are now aware of what goes on. For our performance estimator we should average the randomness out over the repetitions. This makes the estimator better. We should still worry about our data / model, if the randomness before averaging was very high.

**Comment by berndbischl** [comment]
We can now also use statistical tests to compare performance distributions between models, instead of just using mean performance values. 
I will not go further into this, but simply state that in ML with CV, getting the test right for such a comparison is much harder as one might think. This test (if we would have a suitable one...) now guards us against inference which is spurious and simply due to random fluctuation. 

Now lets get back to the "fixing the randomness" issue. Do we now really need that anymore? From a technical perspective, as a programmer, I might sometimes say "yes....".
From a scientific perspective, I don't think so. I don't care if you cannot reproduce EXACTLY my results to the 10th digit, because you used a slightly different CV. 
What I would worry about is when you arrive at SUBSTANTIALLY DIFFERENT INFERENCE than I did, when you do the same thing that I did. If that happens, it means our experimental setup and / or analysis methods are crap.

**Comment by mlindauer** [comment]
Good points and in principle, I agree.
Furthermore, in the long term, I would like to support repetitions of the CV, as it was intended in the format.

In general, I think reproducibility is really important, especially in science.
And the setting of AS induces special attention.
A small variation of the model can induce another selected algorithm,
which can have, e.g., an exponential worse runtime. 
So, depending on the evaluation metric (e.g., PAR10 or mcp),
wrong decisions can heavily worsen the performance of our algorithm selector.
Therefore, I think even slight variations (due to randomness) of the models should be avoided
and the models should be as reproducible as possible.

**Comment by berndbischl** [comment]
> Furthermore, in the long term, I would like to support repetitions of the CV, 
> as it was intended in the format

The arguments in this thread are exactly the reasons why I proposed this in the first place.
It is also only a minor problem and cannot be hard to do (for Lars).

**Comment by berndbischl** [comment]
> In general, I think reproducibility is really important, especially in science.

100% agreed. And it is an even a broader subject than what we are discussion here.

> And the setting of AS induces special attention.
> A small variation of the model can induce another selected algorithm,
> which can have, e.g., an exponential worse runtime. 
> So, depending on the evaluation metric (e.g., PAR10 or mcp),
> wrong decisions can heavily worsen the performance of our algorithm selector.
> Therefore, I think even slight variations (due to randomness) of the models should be avoided
> and the models should be as reproducible as possible.

Actually, IMHO you are mixing two different subjects in the paragraph above. 

a) Reproducibility. For me, this means, being able to (at least qualitatively) to reprodce the results of the study of another scientist.

b) Model stability: I think this is what you worry about here now. It means, that if the data only slightly changes the model (or better: model output?) does not. 
(So we might compare it to something like smoothness in functions)

Now, first of all we would like to DETECT instability. This is exactly what resampling can be used for. Select a slightly different data subset, what happens? Large change? Or you could do other kinds of sensesitivity analysis. But thhrough our evaluation we should be able to see whether this is a problem for a model or not.

Then, if it IS a problem, one might ask how to mitigate it. Regularization is a technique. Or bagging. Or possibly feature reduction. Or ....

 I will write a further comment regarding your original question about what to do if models are stochastic.

**Comment by berndbischl** [comment]
Closing this now, discussion seems to have ended and we do not do anything wrong currently in our setup IMHO.
